src/credentials/secondaryAuth.ts

```diff
@@ -0,0 +1,173 @@
+/*!
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import globals from '../shared/extensionGlobals'
+
+import * as vscode from 'vscode'
+import { getLogger } from '../shared/logger'
+import { showQuickPick } from '../shared/ui/pickerPrompter'
+import { cast, Optional } from '../shared/utilities/typeConstructors'
+import { Auth, Connection } from './auth'
+import { once } from '../shared/utilities/functionUtils'
+import { UnknownError } from '../shared/errors'
+
+async function promptUseNewConnection(newConn: Connection, oldConn: Connection, tools: string[]) {
+    // Multi-select picker would be better ?
+    const saveConnectionItem = {
+        label: `Yes, use ${newConn.label} with ${tools.join(', ')} while using ${oldConn.label} with other services.`,
+        detail: `To remove later, select "Remove Connection from Tool" from the tool's context (right-click) menu.`,
+        data: 'yes',
+    } as const
+
+    const useConnectionItem = {
+        label: `No, switch everything to authenticate with ${newConn.label}.`,
+        detail: 'This will not log you out; you can reconnect at any time by switching connections.',
+        data: 'no',
+    } as const
+
+    const resp = await showQuickPick([saveConnectionItem, useConnectionItem], {
+        title: `Some tools you've been using don't work with ${newConn.label}. Keep using ${newConn.label} in the background while using ${oldConn.label}?`,
+        placeholder: 'Confirm choice',
+    })
+
+    return resp
+}
+
+let oldConn: Auth['activeConnection']
+const auths = new Map<string, SecondaryAuth>()
+const registerAuthListener = once(() => {
+    Auth.instance.onDidChangeActiveConnection(async conn => {
+        const potentialConn = oldConn
+        if (conn !== undefined && potentialConn?.state === 'valid') {
+            const saveableAuths = Array.from(auths.values()).filter(
+                a => !a.isUsingSavedConnection && a.isUsable(potentialConn) && !a.isUsable(conn)
+            )
+            const toolNames = saveableAuths.map(a => a.toolLabel)
+            if (saveableAuths.length > 0 && (await promptUseNewConnection(potentialConn, conn, toolNames)) === 'yes') {
+                await Promise.all(saveableAuths.map(a => a.saveConnection(potentialConn)))
+            }
+        }
+
+        oldConn = conn
+    })
+})
+
+export function getSecondaryAuth<T extends Connection>(
+    toolId: string,
+    toolLabel: string,
+    isValid: (conn: Connection) => conn is T
+): SecondaryAuth<T> {
+    const auth = new SecondaryAuth(toolId, toolLabel, isValid)
+    auths.set(toolId, auth)
+    registerAuthListener()
```

Comment:
off-topic: could once() work at the call-site instead of declaration? Would be much clearer I think.

```suggestion
    once(registerAuthListener)
```

Reply:
> off-topic: could once() work at the call-site instead of declaration? Would be much clearer I think.

Yeah it can be done although the behavior would still be associated with the function rather than the callsite. So like `once(foo)` would only invoke `foo` once regardless of where you called `once`.